### PR TITLE
Implement majority of more efficient Redshift Copy Strategy.

### PR DESCRIPTION
**Pull request by davinchia**
## What
Solves the redshift portion of #1958.

Technical approach discussed [here](https://docs.google.com/document/d/1BeDx5v4t9JcO2PVUkV7iXlxgro3AuTc4JrnUiqssHfg/edit#heading=h.gn4ad5et1brq).

## How
TLDR:
Instead of inserts, we write the data to S3 and issue a COPY command copying from S3 into redshift.

- Use a single file as its sufficiently performant and we do not want to introduce file destination related operations yet.
- Use an open source library for uploads as AWS does not natively support streaming loads.

See the linked doc for more detail.

My intention with this PR is to first implement the meat of the write-and-copy strategy. This is mainly centered around the `RedshiftCopier` class. I plan hook up the `RedshiftCopier` to the actual Destination class, and implement all the plumbing, in a follow up PR.

## Pre-merge Checklist
- [ ] *Finish speed tests.*

## Recommended reading order
1. `RedshiftCopier`
2. `RedshiftCopyDestination`

### CODE REVIEW

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@Disabled("Disabled on automatic execution as the tests requires specific credentials to run."
+    + "Intended to be a lighter-weight complement to the Docker integration tests. Fill in the appropriate blank variables to run.")
+public class RedshiftCopierTest {
+
+  // Fill in the empty variables before running.
+  private static final String S3_KEY_ID = "";
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java:55]: will switch this to an env file before merging.

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  private static final String REDSHIFT_CONNECTION_STRING = "";
+  private static final String REDSHIFT_USER = "";
+  private static final String REDSHIFT_PASS = "";
+
+  private static final String S3_REGION = "us-west-2";
+  private static final String TEST_BUCKET = "redshift-copier-test";
+  private static final String RUN_FOLDER = "test-folder";
+  private static final String SCHEMA_NAME = "public";
+  private static final String STREAM_NAME = "redshift_copier_test";
+  private static final String RAW_TABLE_NAME = new RedshiftSQLNameTransformer().getRawTableName(STREAM_NAME);
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java:66]: even though `getRawTableName` is deprecated - it's less confusing to use the deprecated method for now, and refactor everything at one go later. I'll put up a PR once everything is done.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/DestinationConsumer.java [hunk 3]
```diff
@@ -26,12 +26,21 @@
...
+ *
+ * This is via the accept method, which commonly handles parsing, validation, batching and writing
+ * of the transformed data to the final destination i.e. the technical system data is being written
+ * to.
+ *
+ * Lifecycle: 1. Instantiate consumer. 2. start() to initialize any resources that need to be
+ * created BEFORE the consumer consumes any messages. 3. Consumes ALL records via
+ * {@link DestinationConsumer#accept(T)} 4. Always (on success or failure) finalize by calling
+ * {@link DestinationConsumer#close()}
+ *
```
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/DestinationConsumer.java:41]: Nice, too bad the spotlessStyling is messing up the list view...

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+      multipartUploadManager.abort();
+      return;
+    }
+    closeS3WriteStreamAndUpload();
+
+    var tmpTableName = NAMING_RESOLVER.getTmpTableName(streamName);
+    var sqlOp = new RedshiftSqlOperations();
+    try {
+      createTmpTableAndCopyS3FileInto(tmpTableName, sqlOp);
+      mergeIntoDestTableIncrementalOrFullRefresh(tmpTableName, sqlOp);
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:154]: is there a way to do this in a single transaction for all tables? if not then we need to document expectations in the error cases. currently for other destinations, we are able to do this copy atomically.

**Review by cgardens — CHANGES_REQUESTED** [review]
the approach makes sense. is the goal to merge this and then do a separate PR or to keep this open until your second PR is done? the bones of this seem right, one concern about the atomicity of the copy operation. after that's resolved i think we will be good.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java [hunk 5]
```diff
@@ -73,4 +73,24 @@ public void insertRecords(JdbcDatabase database, Stream<AirbyteRecordMessage> re
...
+                                     String s3FileLocation,
+                                     String schema,
+                                     String tableName,
+                                     String s3KeyId,
+                                     String s3Key,
+                                     String s3Region)
+      throws SQLException {
+    final var copyQuery = String.format(
+        "COPY %s.%s FROM '%s'\n"
+            + "CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'\n"
```
[the remark this answers is not in the document]
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java:86]: You are right. The more recommended way is to use roles. Will change.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopyDestination.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+      client.createBucket(createBucketRequest);
+    }
+  }
+
+  private static class RedshiftCopyDestinationConsumer extends FailureTrackingConsumer<AirbyteMessage> {
+
+    private final AmazonS3 client;
+    private final ConfiguredAirbyteCatalog catalog;
+    private final String redshiftRegion;
+    private Map<String, RedshiftCopier> tableNameToCopier;
```
[the remark this answers is not in the document]
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopyDestination.java:124]: Yes it should. I've left it out for now; planned for these changes to be part of the follow up PR.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    this.syncMode = syncMode;
+    this.schemaName = schema;
+    this.streamName = streamName;
+    this.s3Client = client;
+    this.database = database;
+    this.s3KeyId = s3KeyId;
+    this.s3Key = s3key;
+    this.s3Region = s3Region;
+
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
```
[the remark this answers is not in the document]
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:115]: Yes if we set too high of a queue * part limit combination. As of now, our memory consumption should 200 MB (100MB sitting in queue, 100MB uploading), which seems okay for now? I can leave a comment warning devs off setting this too high?

File: airbyte-integrations/connectors/destination-redshift/build.gradle [hunk 8]
```diff
@@ -19,17 +19,19 @@ dependencies {
...
+    testImplementation 'io.findify:s3mock_2.13:0.2.6'
     testImplementation 'org.apache.commons:commons-text:1.9'
     testImplementation 'org.apache.commons:commons-lang3:3.11'
     testImplementation 'org.apache.commons:commons-dbcp2:2.7.0'
 
-    testImplementation project(':airbyte-test-utils')
-
-    integrationTestJavaImplementation project(':airbyte-integrations:bases:standard-destination-test')
-
-    implementation files(project(':airbyte-integrations:bases:base-java').airbyteDocker.outputs)
```
[the remark this answers is not in the document]
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/build.gradle:33]: haha I was talking to someone and didn't realise we have a convention. Will chat with you about this during our sync later.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java (hunk 4 above)
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:154]: great point. nothing jumps out at me for now; will play around and change this.

**Comment by davinchia** [comment]
@cgardens thanks for the quick turnaround. plan is to merge this in first and open another PR. want to not leave hanging PRs and I myself don't like merging in massive blocks of code. lmk if we generally don't favor that. will work on the feedback now

**Review by cgardens — APPROVED** [review]
@davinchia and I discussed the transaction piece. Switching to approve because we are the same page on what needs to be done and I don't block this PR due to timezone concerns.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    // See the above comment.
+    this.outputStream = multipartUploadManager.getMultiPartOutputStreams().get(0);
+
+    var writer = new PrintWriter(outputStream, true, StandardCharsets.UTF_8);
+    this.csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT);
+  }
+
+  public static void closeAsOneTransaction(List<RedshiftCopier> copiers, boolean hasFailed, JdbcDatabase redshiftDb) throws Exception {
+    try {
+      StringBuilder mergeCopiersToFinalTableQuery = new StringBuilder();
```
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:140]: @cgardens this was the cleanest way I could find of doing this - basically the same pattern as the `JdbcBufferedConsumerFactory`. If I understand correctly, the `executeTransaction` call on L145 will automatically rollback any inserts into the final destination tables if inserts run into any errors.

The tmp table and the S3 files will be cleaned up by the finally block.

**Pull request merged**